source/opt/aggressive_dead_code_elim_pass.cpp

```diff
@@ -0,0 +1,504 @@
+  }
+  for (auto& ai : module_->annotations()) {
+    if (ai.opcode() != SpvOpDecorate && ai.opcode() != SpvOpDecorateId)
+      continue;
+    DeleteInstIfTargetDead(&ai);
+    modified = true;
+  }
+  // Kill dead instructions
+  for (auto& blk : *func) {
+    for (auto& inst : blk) {
+      if (dead_insts_.find(&inst) == dead_insts_.end())
+        continue;
+      def_use_mgr_->KillInst(&inst);
+      modified = true;
+    }
+  }
+  return modified;
+}
+
+void AggressiveDCEPass::Initialize(ir::Module* module) {
+
+  module_ = module;
+
+  // Initialize id-to-function map
+  id2function_.clear();
+  for (auto& fn : *module_)
+    id2function_[fn.result_id()] = &fn;
+
+  def_use_mgr_.reset(new analysis::DefUseManager(consumer(), module_));
+}
+
+Pass::Status AggressiveDCEPass::ProcessImpl() {
+  // Current functionality assumes shader capability 
+  // TODO(greg-lunarg): Handle additional capabilities
+  if (!module_->HasCapability(SpvCapabilityShader))
+    return Status::SuccessWithoutChange;
+
+  // Current functionality assumes logical addressing only
+  // TODO(greg-lunarg): Handle non-logical addressing
+  if (module_->HasCapability(SpvCapabilityAddresses))
+    return Status::SuccessWithoutChange;
+
+  // If any extensions in the module are not explicitly allowed,
+  // return unmodified. Currently, no extensions are allowed.
+  // glsl_std_450 extended instructions are allowed.
+  // TODO(greg-lunarg): Allow additional extensions
+  if (!AllExtensionsAllowed())
+    return Status::SuccessWithoutChange;
+
+  InitCombinatorSets();
+
+  bool modified = false;
+  for (auto& e : module_->entry_points()) {
+    ir::Function* fn =
+        id2function_[e.GetSingleWordInOperand(kEntryPointFunctionIdInIdx)];
+    modified = AggressiveDCE(fn) || modified;
+  }
+  return modified ? Status::SuccessWithChange : Status::SuccessWithoutChange;
+}
+
+AggressiveDCEPass::AggressiveDCEPass()
+    : module_(nullptr), def_use_mgr_(nullptr) {}
+
+Pass::Status AggressiveDCEPass::Process(ir::Module* module) {
+  Initialize(module);
+  return ProcessImpl();
+}
+
+void AggressiveDCEPass::InitCombinatorSets() {
+  combinator_ops_shader_ = {
+    SpvOpNop,
+    SpvOpUndef,
+    SpvOpVariable,
+    SpvOpImageTexelPointer,
+    SpvOpLoad,
+    SpvOpAccessChain,
+    SpvOpInBoundsAccessChain,
+    SpvOpArrayLength,
+    SpvOpVectorExtractDynamic,
+    SpvOpVectorInsertDynamic,
+    SpvOpVectorShuffle,
+    SpvOpCompositeConstruct,
+    SpvOpCompositeExtract,
+    SpvOpCompositeInsert,
+    SpvOpCopyObject,
+    SpvOpTranspose,
+    SpvOpSampledImage,
+    SpvOpImageSampleImplicitLod,
+    SpvOpImageSampleExplicitLod,
+    SpvOpImageSampleDrefImplicitLod,
+    SpvOpImageSampleDrefExplicitLod,
+    SpvOpImageSampleProjImplicitLod,
+    SpvOpImageSampleProjExplicitLod,
+    SpvOpImageSampleProjDrefImplicitLod,
+    SpvOpImageSampleProjDrefExplicitLod,
+    SpvOpImageFetch,
+    SpvOpImageGather,
+    SpvOpImageDrefGather,
+    SpvOpImageRead,
+    SpvOpImage,
```

Comment:
Could add instructions enabled by the ImageQuerycapability.
Optional: Add TODO mentioning these.